### ISSUE TITLE
Transactions appear to leak data between sessions, resulting in EntityNotFoundException

### ISSUE THREAD

**Issue by subvertallchris** [neo4j/neo4j#3864]
Tested and confirmed with Neo4j 2.1.6, 2.2.0-M01, 2.2.0-M02 using Neo4j.rb 4.1.0 and Neo4j-core 4.0.0 Ruby gems.

Given two completely independent sessions, queries against the Cypher transactional appear to be leaking IDs, resulting in errors in one session that reference nodes in the other.

In session 1:

```
Benchmark.ips do |x|
  x.config(warmup: 3, time: 20)
  x.report('create/destroy') do
    s = Student.create
    s.destroy
  end
end
```
`Benchmark.ips` runs the given commands as many times as it can for the duration given.

The queries against the server are `CREATE (s:Student { uuid: {uuid} ) RETURN s` and `MATCH (s) WHERE ID(s) = {id_of_s} OPTIONAL MATCH s-[r]-() DELETE s,r`. Each is an independent transaction. Four queries total: cypher query then commit to create, cypher query then commit to destroy. Uniqueness on the Student's uuid is enforced through a constraint.

In session 2:
```
# Manually, repeatedly
Student.count
```
Session 2 will occasionally eventually return an `org.neo4j.cypher.EntityNotFoundException` error referencing Node with ID 429, which is a node created by session **1**. 

This appears to be the same issue as http://stackoverflow.com/questions/26133131/are-neo4js-cypher-transactions-broken. The poster said they would open an issue but if they did, I couldn't find it.

**Comment by subvertallchris** [comment]
Session 2, which is just doing `MATCH (s:Student) RETURN COUNT(s)`, is running against the legacy Cypher endpoint and not the cypher transaction endpoint.

I also just found that this error occurs when I perform that same query in the Neo4j shell.
```
neo4j-sh (?)$ PROFILE MATCH (s:Student) RETURN COUNT(s);
EntityNotFoundException: Node with id 2579
```

**Comment by bunkat** [comment]
I think I've hit something similar in the past, see http://stackoverflow.com/questions/24442142/neo4j-2-1-2-unable-to-delete-nodes-using-cypher-query-that-worked-with-neo4j-2.  Deleting nodes threw `Entity Not Found` exceptions when there shouldn't have been a way to generate that exception since `DELETE` doesn't throw if a node doesn't exist.

I filed https://github.com/neo4j/neo4j/issues/2122 which never got any response and is still open.

**Comment by subvertallchris** [comment]
This is really disturbing. Every action in Neo4j.rb occurs within a transaction, so everyone using the driver is at high risk for this.

I pinged @jexp about it last night and he said he'll take a look, so I feel a bit more comfortable that we'll get a resolution soon.

**Comment by subvertallchris** [comment]
I made a repo containing a simple Rails app with one model and two class methods to easily reproduce. https://github.com/subvertallchris/debugtest

**Comment by subvertallchris** [comment]
Another update: swapping out the node IDs for app-generated UUIDs in the `DELETE` statement doesn't fix the issue.

**Comment by chrisvest** [comment]
Does it help if you set `cache_type=none` in neo4j.properties?

**Comment by subvertallchris** [comment]
No, it doesn't.

**Comment by subvertallchris** [comment]
Just fixed a typo in my initial post. It should have said, "Session 1 will occasionally eventually return an org.neo4j.cypher.EntityNotFoundException error referencing Node with ID 429, which is a node created by session 2."

**Comment by cdshines** [comment]
I concur. This is perhaps related to the issue I opened a couple of months ago. I managed to overcome the problem by restarting transactions that throw `TransactionFailureException`, causes of which are `EntityNotFoundExcepion`s, but that's hardly a nice solution.

**Comment by subvertallchris** [comment]
Could you link to that issue? Would be cool to have one thread tracking all of the reports. I'd also love to see the code where you recover and restart, if that's online. I might need to figure out a kludge for the Ruby gem, though I'm not sure if it'll be possible.

**Comment by cdshines** [comment]
@subvertallchris, it's #3489. The explanation is rather weak, but i still have exception logs coming to me from users from time to time. I'll update this thread with them in case I get another one.

BTW, I'm using an embedded version and here's one of the things I tried. It helped but I'm unsure about it's reliability, so I don't use it in real life (for now maybe).

	public static Object getProperty(PropertyContainer pc, String key, boolean isMandatory) {
		Object value;
		try {
			value = pc.getProperty(key);
		} catch (NotFoundException e) {
			if (isMandatory) {
				if (e.getCause() instanceof PropertyNotFoundException) {
					return getProperty(pc, key, true);
				} else {
					throw e;
				}
			} else {
				throw e;
			}
		}
		return value;
	}

**Comment by bunkat** [comment]
This is what I've had to do as well.  I retry both `NotFoundException` and `DeadlockDetected` exceptions, waiting longer on each retry in hopes that a window opens up when they will succeed. Even doing this, I still see transactions outright fail when using Cypher over REST without returning any exception.

**Comment by subvertallchris** [comment]
Are you guys storing each query so you can replay everything in order when this bug pops up, or does the Java API provide some sort of `TransactionReplay` function?

**Comment by bunkat** [comment]
Not sure about the Java API. I'm using Cypher with Go and I store all of the queries and their parameters in my own transaction object.  I can then just retry the my transaction as needed (or dump the whole thing for debugging purposes).

**Comment by cdshines** [comment]
I'm using Java API and have to do it in a banal loop with exception checks. And no, there aren't any replay facilities. But since we have a distributed system with neo4j playing more than one role, catching failing transaction here and there and trying to stay consistent is quite a pain.

**Comment by subvertallchris** [comment]
Didn't think so, that's what I was thinking of doing in Neo4j.rb if a kludge is needed. This feels like a pretty serious bug, though, so I'm keeping my fingers crossed that Neo can find a solution now that it can be easily reproduced.

**Comment by bunkat** [comment]
Very much appreciated @subvertallchris for finding a way to reproduce it.

**Comment by mneedham** [comment]
Hi @subvertallchris,

Thanks for linking to your repo. I've got that running on my machine and I am seeing the same behaviour as you. 

I'm trying to write an equivalent test that just uses the underlying Java API to try and narrow down what's going out. Haven't got it to fail yet but will keep at it.

Cheers
Mark

**Comment by subvertallchris** [comment]
Awesome, that sounds like a first step, at least. I added a note to my initial post just now, I had forgotten to mention that the `Student` label has a uniqueness constraint on its `uuid` property, which is generated by the app prior to creation. Its `CREATE` is more like `CREATE (s:Student { uuid: {uuid} }) RETURN s`.

**Comment by subvertallchris** [comment]
Hey @mneedham, any luck with this?

**Comment by mneedham** [comment]
Hi, 

So we've looked into this and been able to reproduce. It's quite clear that transactions are committed (and their results become visible to other transactions) while the query is iterating through the collection of nodes with a given label. So the collection can contain nodes that existed when the query started, but have been deleted during the course of running the query. 

There is a problem with the way Cypher handles iterating over a collection that can change while the query is executing.  We'll get this fixed in a future release.

In the mean time the best workaround would be to put the count query in a retry loop and ignore cases where it tries to count simultaneously deleted nodes.

Cheers
Mark

**Comment by subvertallchris** [comment]
Cool, really glad that some progress has been made. I'm seeing the same symptoms from other actions, too. Is it possible that unique property constraints and/or `CREATE UNIQUE` functions within Cypher are susceptible to the same problem? I can reproduce it consistently but haven't yet taken the time to narrow it down to one or the other. Can share some example code if it'd be helpful to you.

**Comment by mneedham** [comment]
@subvertallchris I'm not quite sure I understand so yeh example code would be useful.

**Comment by subvertallchris** [comment]
Yeah, I thought that might come off as kind of vague...

I just pushed to a new branch on my repo, `v2`, that demonstrates what I'm talking about. It uses a new branch on neo4j-core to give you slightly better output from the error. Clone it and `bundle update` to grab the new branch of neo4j-core.

There are two new methods in the debugtest v2 branch: `Student.rapid_rels` and `Student.rapid_rels_transaction`. Each does the following:

* Creates a Student and Lesson node
* Creates a relationship between them
* Matches that rel and returns one node
* Deletes both nodes

`rapid_rels` does it in a series of small transactions, as in the last test. Each transaction is a two parter: perform the action and leave the transaction open, then close the transaction

`rapid_rels_transaction` does one big transaction with all actions nested inside of it.

```ruby
  def self.rapid_rels
    Benchmark.ips do |x|
      x.config(warmup: 2, time: 20)
      x.report('create, create, rel, destroy') do
        s = Student.create(name: 'Lauren'); l = Lesson.create(subject: 'Math'); s.lessons << l; s.lessons.first; s.destroy; l.destroy
      end
    end
  end

  def self.rapid_rels_transaction
    Benchmark.ips do |x|
      x.config(warmup: 2, time: 20)

      x.report('transaction create, create, rel, destroy') do
        begin
          tx = Neo4j::Transaction.new
          s = Student.create(name: 'Lauren'); l = Lesson.create(subject: 'Math'); s.lessons << l; s.lessons.first; s.destroy; l.destroy
        ensure
          tx.close
        end
      end
    end
  end
```

The `Lesson` label has a uniqueness constraint on the `uuid` property, just like `Student.

What I'm finding is that with two independent sessions, if we run each command in one window, we'll crash with either the same `EntityNotFoundException` or `Unrecognized transaction id. Transaction may have timed out and been rolled back.` It doesn't happen if both windows run the same command... I'm guessing it has something to do with the one opening and closing transactions quickly while the other leaves a transaction open, so there's more change of a conflict? The `Unrecognized transaction id` error happens significantly more often than the EntityNotFoundException.

**Comment by mneedham** [comment]
Hi @subvertallchris - sorry for the delayed reply, was at FOSDEM over the weekend. Going to checkout your branch and look into this now.

**Comment by subvertallchris** [comment]
NP, good luck. I'm signed into Skype as chrislgrigg if you want me to test anything.

**Comment by mneedham** [comment]
@subvertallchris I grabbed the 'part2' branch from your repository - I assume that's the one I was supposed to grab?

Then `rails c` on two terminals and I ran this on one:

```
Student.rapid_rels
```

And this on the other:

```
Student.rapid_rels_transaction
```

But everything was happy. I tried on 2.1.6 and 2.2.0-M03. Can you attach the data/graph.db/messages.log & data/log/*.log files for where you see it failing?

**Comment by subvertallchris** [comment]
Yup, that's the one. Weird, it happens to me pretty quickly.

console.log: https://gist.github.com/subvertallchris/280f06b9cf6db7a4a88f
neo4j.0.0.log: https://gist.github.com/subvertallchris/c243a88cfd518ef56f7d
messages.log: https://gist.github.com/subvertallchris/289adcf183dabb3c5496

**Comment by mneedham** [comment]
Hmmm there's nothing in there, that's weird - exceptions usually make their way into one of those files. Do you just see the exception in your console then?

**Comment by subvertallchris** [comment]
That's right. Is there a way to turn up the log level?

**Comment by subvertallchris** [comment]
Also, did you try running it a few times? I got through one iteration just now that didn't explode.

**Comment by mneedham** [comment]
Yeh I've run it 5 or 6 times and it didn't break as yet. Can you paste the full exception for the two different ones you've been seeing?

**Comment by subvertallchris** [comment]
From what I'm able to see, it seems to happen most frequently right after I start the server, before cache is totally warmed up. After that, it becomes harder to reproduce. That might just be in my head, though...

Here's the first one, which happens the majority of the time. Unfortunately, I'm working with the output from the neo4j-core gem, so I'll have to modify it to get more info from the server's response.

```
 Student#lessons 22ms MATCH (student10696:`Student`), (result_lessons:`Lesson`), student10696-[rel0:`ENROLLED_IN`]->(result_lessons:`Lesson`) WHERE ID(student10696) = {ID_student10696} RETURN result_lessons ORDER BY ID(result_lessons) LIMIT {limit_1} | {:ID_student10696=>10696, "limit_1"=>1}
Neo4j::Server::Resource::ServerException: Expected response code 200 Error for request, body: {"results"=>[], "errors"=>[{"code"=>"Neo.ClientError.Transaction.UnknownId", "message"=>"Unrecognized transaction id. Transaction may have timed out and been rolled back."}]} http://localhost:7474/db/data/transaction/3550/commit, 404, 404
```

**Comment by mneedham** [comment]
@subvertallchris  you can up the log level via this file - https://github.com/neo4j/neo4j/blob/master/community/kernel/src/main/resources/neo4j-logback.xml e.g. change it to DEBUG

I'm not sure why you're seeing that exception though. Neo4j internally keeps track of currently running transactions in a ConcurrentHashMap but somehow in this case it tries to rollback one which isn't in the map...which means it's probably been removed by another thread somewhere. 

There is a background thread which runs on a scheduler and cleans up suspended transactions so I wonder if that's what's happening. Then the question becomes why is it getting suspended. 

If I can get it to happen I can instrument the code and figure out what's going on. 

What's also weird is that looking at the code around this area there are messages logged at the info & error levels all over the place so I'm not sure why nothing is showing up in any of your log files...

**Comment by subvertallchris** [comment]
Amateur question here, where would I find or put that file? I see other logbacks in `conf`, do I need to drop it in there and register it somehow?

**Comment by mneedham** [comment]
There should be one named neo4j-logback.xml there already hopefully that you can edit. If not then you can just create it. It should be automatically picked up but I guess you can check by seeing if `DEBUG` lines start showing up in the log files

**Comment by subvertallchris** [comment]
OK, got it working. It wasn't picking up the new logback but I dropped its contents into the custom-logback.xml and it started working. Give me a few mins to see if I can get it to log the error.

**Comment by subvertallchris** [comment]
Tada!

```
2015-02-02 17:54:38.301+0000 DEBUG [o.n.k.i.q.QueryExecutionEngine]: MATCH (n) WHERE ID(n) = {neo_id} OPTIONAL MATCH n-[r]-() DELETE n, r
2015-02-02 17:54:38.301+0000 DEBUG [o.n.k.i.q.QueryExecutionEngine]: MATCH (n) WHERE ID(n) = {neo_id} OPTIONAL MATCH n-[r]-() DELETE n, r
2015-02-02 17:54:38.311+0000 DEBUG [o.n.k.i.q.QueryExecutionEngine]: MATCH (n) WHERE ID(n) = {neo_id} OPTIONAL MATCH n-[r]-() DELETE n, r
2015-02-02 17:54:38.311+0000 DEBUG [o.n.k.i.q.QueryExecutionEngine]: MATCH (n) WHERE ID(n) = {neo_id} OPTIONAL MATCH n-[r]-() DELETE n, r
2015-02-02 17:54:38.317+0000 DEBUG [o.n.k.i.q.QueryExecutionEngine]: CREATE (n:`Student` {props}) RETURN ID(n)
2015-02-02 17:54:38.317+0000 DEBUG [o.n.k.i.q.QueryExecutionEngine]: CREATE (n:`Student` {props}) RETURN ID(n)
2015-02-02 17:54:38.323+0000 ERROR [o.n.s.r.t.TransactionFacade]: Failed to commit transaction.
java.lang.RuntimeException: org.neo4j.kernel.api.exceptions.TransactionFailureException: Transaction rolled back even if marked as successful
	at org.neo4j.server.rest.transactional.TransitionalTxManagementKernelTransaction.commit(TransitionalTxManagementKernelTransaction.java:87) ~[neo4j-server-2.2.0-M03.jar:2.2.0-M03]
	at org.neo4j.server.rest.transactional.TransactionHandle.closeContextAndCollectErrors(TransactionHandle.java:277) [neo4j-server-2.2.0-M03.jar:2.2.0-M03]
	at org.neo4j.server.rest.transactional.TransactionHandle.commit(TransactionHandle.java:146) [neo4j-server-2.2.0-M03.jar:2.2.0-M03]
	at org.neo4j.server.rest.web.TransactionalService$2.write(TransactionalService.java:210) ~[neo4j-server-2.2.0-M03.jar:2.2.0-M03]
	at com.sun.jersey.core.impl.provider.entity.StreamingOutputProvider.writeTo(StreamingOutputProvider.java:71) ~[jersey-core-1.18.1.jar:1.18.1]
	at com.sun.jersey.core.impl.provider.entity.StreamingOutputProvider.writeTo(StreamingOutputProvider.java:57) ~[jersey-core-1.18.1.jar:1.18.1]
	at com.sun.jersey.spi.container.ContainerResponse.write(ContainerResponse.java:302) ~[jersey-server-1.18.1.jar:1.18.1]
	at com.sun.jersey.server.impl.application.WebApplicationImpl._handleRequest(WebApplicationImpl.java:1510) ~[jersey-server-1.18.1.jar:1.18.1]
	at com.sun.jersey.server.impl.application.WebApplicationImpl.handleRequest(WebApplicationImpl.java:1419) ~[jersey-server-1.18.1.jar:1.18.1]
	at com.sun.jersey.server.impl.application.WebApplicationImpl.handleRequest(WebApplicationImpl.java:1409) ~[jersey-server-1.18.1.jar:1.18.1]
	at com.sun.jersey.spi.container.servlet.WebComponent.service(WebComponent.java:409) ~[jersey-servlet-1.18.1.jar:1.18.1]
	at com.sun.jersey.spi.container.servlet.ServletContainer.service(ServletContainer.java:540) ~[jersey-servlet-1.18.1.jar:1.18.1]
	at com.sun.jersey.spi.container.servlet.ServletContainer.service(ServletContainer.java:715) ~[jersey-servlet-1.18.1.jar:1.18.1]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790) ~[javax.servlet-api-3.1.0.jar:3.1.0]
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:751) ~[jetty-servlet-9.2.1.v20140609.jar:9.2.1.v20140609]
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:566) ~[jetty-servlet-9.2.1.v20140609.jar:9.2.1.v20140609]
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:219) ~[jetty-server-9.2.1.v20140609.jar:9.2.1.v20140609]
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1111) ~[jetty-server-9.2.1.v20140609.jar:9.2.1.v20140609]
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:498) ~[jetty-servlet-9.2.1.v20140609.jar:9.2.1.v20140609]
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:183) ~[jetty-server-9.2.1.v20140609.jar:9.2.1.v20140609]
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1045) ~[jetty-server-9.2.1.v20140609.jar:9.2.1.v20140609]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141) ~[jetty-server-9.2.1.v20140609.jar:9.2.1.v20140609]
	at org.eclipse.jetty.server.handler.HandlerList.handle(HandlerList.java:52) ~[jetty-server-9.2.1.v20140609.jar:9.2.1.v20140609]
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:98) ~[jetty-server-9.2.1.v20140609.jar:9.2.1.v20140609]
	at org.eclipse.jetty.server.Server.handle(Server.java:461) ~[jetty-server-9.2.1.v20140609.jar:9.2.1.v20140609]
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:284) ~[jetty-server-9.2.1.v20140609.jar:9.2.1.v20140609]
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:244) ~[jetty-server-9.2.1.v20140609.jar:9.2.1.v20140609]
	at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:534) ~[jetty-io-9.2.1.v20140609.jar:9.2.1.v20140609]
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:607) ~[jetty-util-9.2.1.v20140609.jar:9.2.1.v20140609]
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:536) ~[jetty-util-9.2.1.v20140609.jar:9.2.1.v20140609]
	at java.lang.Thread.run(Thread.java:724) ~[na:1.7.0_25]
Caused by: org.neo4j.kernel.api.exceptions.TransactionFailureException: Transaction rolled back even if marked as successful
	at org.neo4j.kernel.impl.api.KernelTransactionImplementation.close(KernelTransactionImplementation.java:401) ~[neo4j-kernel-2.2.0-M03.jar:2.2.0-M03]
	at org.neo4j.server.rest.transactional.TransitionalTxManagementKernelTransaction.commit(TransitionalTxManagementKernelTransaction.java:83) ~[neo4j-server-2.2.0-M03.jar:2.2.0-M03]
	... 30 common frames omitted
2015-02-02 17:54:38.323+0000 ERROR [o.n.s.r.t.TransactionFacade]: Failed to commit transaction.
java.lang.RuntimeException: org.neo4j.kernel.api.exceptions.TransactionFailureException: Transaction rolled back even if marked as successful
	at org.neo4j.server.rest.transactional.TransitionalTxManagementKernelTransaction.commit(TransitionalTxManagementKernelTransaction.java:87) ~[neo4j-server-2.2.0-M03.jar:2.2.0-M03]
	at org.neo4j.server.rest.transactional.TransactionHandle.closeContextAndCollectErrors(TransactionHandle.java:277) [neo4j-server-2.2.0-M03.jar:2.2.0-M03]
	at org.neo4j.server.rest.transactional.TransactionHandle.commit(TransactionHandle.java:146) [neo4j-server-2.2.0-M03.jar:2.2.0-M03]
	at org.neo4j.server.rest.web.TransactionalService$2.write(TransactionalService.java:210) ~[neo4j-server-2.2.0-M03.jar:2.2.0-M03]
	at com.sun.jersey.core.impl.provider.entity.StreamingOutputProvider.writeTo(StreamingOutputProvider.java:71) ~[jersey-core-1.18.1.jar:1.18.1]
	at com.sun.jersey.core.impl.provider.entity.StreamingOutputProvider.writeTo(StreamingOutputProvider.java:57) ~[jersey-core-1.18.1.jar:1.18.1]
	at com.sun.jersey.spi.container.ContainerResponse.write(ContainerResponse.java:302) ~[jersey-server-1.18.1.jar:1.18.1]
	at com.sun.jersey.server.impl.application.WebApplicationImpl._handleRequest(WebApplicationImpl.java:1510) ~[jersey-server-1.18.1.jar:1.18.1]
	at com.sun.jersey.server.impl.application.WebApplicationImpl.handleRequest(WebApplicationImpl.java:1419) ~[jersey-server-1.18.1.jar:1.18.1]
	at com.sun.jersey.server.impl.application.WebApplicationImpl.handleRequest(WebApplicationImpl.java:1409) ~[jersey-server-1.18.1.jar:1.18.1]
	at com.sun.jersey.spi.container.servlet.WebComponent.service(WebComponent.java:409) ~[jersey-servlet-1.18.1.jar:1.18.1]
	at com.sun.jersey.spi.container.servlet.ServletContainer.service(ServletContainer.java:540) ~[jersey-servlet-1.18.1.jar:1.18.1]
	at com.sun.jersey.spi.container.servlet.ServletContainer.service(ServletContainer.java:715) ~[jersey-servlet-1.18.1.jar:1.18.1]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790) ~[javax.servlet-api-3.1.0.jar:3.1.0]
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:751) ~[jetty-servlet-9.2.1.v20140609.jar:9.2.1.v20140609]
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:566) ~[jetty-servlet-9.2.1.v20140609.jar:9.2.1.v20140609]
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:219) ~[jetty-server-9.2.1.v20140609.jar:9.2.1.v20140609]
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1111) ~[jetty-server-9.2.1.v20140609.jar:9.2.1.v20140609]
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:498) ~[jetty-servlet-9.2.1.v20140609.jar:9.2.1.v20140609]
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:183) ~[jetty-server-9.2.1.v20140609.jar:9.2.1.v20140609]
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1045) ~[jetty-server-9.2.1.v20140609.jar:9.2.1.v20140609]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141) ~[jetty-server-9.2.1.v20140609.jar:9.2.1.v20140609]
	at org.eclipse.jetty.server.handler.HandlerList.handle(HandlerList.java:52) ~[jetty-server-9.2.1.v20140609.jar:9.2.1.v20140609]
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:98) ~[jetty-server-9.2.1.v20140609.jar:9.2.1.v20140609]
	at org.eclipse.jetty.server.Server.handle(Server.java:461) ~[jetty-server-9.2.1.v20140609.jar:9.2.1.v20140609]
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:284) ~[jetty-server-9.2.1.v20140609.jar:9.2.1.v20140609]
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:244) ~[jetty-server-9.2.1.v20140609.jar:9.2.1.v20140609]
	at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:534) ~[jetty-io-9.2.1.v20140609.jar:9.2.1.v20140609]
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:607) ~[jetty-util-9.2.1.v20140609.jar:9.2.1.v20140609]
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:536) ~[jetty-util-9.2.1.v20140609.jar:9.2.1.v20140609]
	at java.lang.Thread.run(Thread.java:724) ~[na:1.7.0_25]
Caused by: org.neo4j.kernel.api.exceptions.TransactionFailureException: Transaction rolled back even if marked as successful
	at org.neo4j.kernel.impl.api.KernelTransactionImplementation.close(KernelTransactionImplementation.java:401) ~[neo4j-kernel-2.2.0-M03.jar:2.2.0-M03]
	at org.neo4j.server.rest.transactional.TransitionalTxManagementKernelTransaction.commit(TransitionalTxManagementKernelTransaction.java:83) ~[neo4j-server-2.2.0-M03.jar:2.2.0-M03]
	... 30 common frames omitted
2015-02-02 17:54:38.330+0000 DEBUG [o.n.k.i.q.QueryExecutionEngine]: CREATE (n:`Lesson` {props}) RETURN ID(n)
```

**Comment by mneedham** [comment]
Cool. Are you able to tally up the transactions that you get that exception for with what the query actually being executed is? 

I wonder if some transactions are waiting around so long for the lock on the 'Lauren' and 'Math' nodes which you're repeatedly creating and deleting that they eventually timeout? 

I don't think this error is the same type as the original one that you posted though.

**Comment by subvertallchris** [comment]
Sorry, completely missed this comment. Can you elaborate on "Are you able to tally up the transactions that you get that exception for with what the query actually being executed is?"

**Comment by mneedham** [comment]
As in can you see which transaction it is that you got the exception from? Then we can have a  look at other transactions around it which committed (perhaps touching the same data?) and figure out why the error's happening.

**Comment by fdietze** [comment]
This seems to be related: https://github.com/neo4j/neo4j/issues/2677

**Comment by bunkat** [comment]
Any update on this?  I noticed that Neo4j 2.2 has hit RC but this issue hasn't been resolved yet.

**Comment by fdietze** [comment]
My isolation tests(https://github.com/neo4j/neo4j/issues/2677#issuecomment-73754513) seem to be fixed in 2.2 (not in 2.1.7), but I didn't see any reaction form the Neo4j Team on this yet.

**Comment by subvertallchris** [comment]
That sounds promising. I haven't had a chance to test, stuck in 2.1.7 and just keeping my eyes on the prize lately for work. Anyone else able to confirm?

**Comment by subvertallchris** [comment]
Just tested with 2.2.0, still occurs.

**Comment by bunkat** [comment]
Any update on this from the Neo4j team?

**Comment by jeppebemad** [comment]
Also experiencing this. Very frustrating! 
Confirmed in 2.2.3 and 2.2.5.

**Comment by EmergentBehavior** [comment]
Same here, though in Go.

**Comment by jakewins** [comment]
Super sorry for the slow pace we've been on for this. The #2677 is an unrelated issue which is indeed resolved in 2.2. This issue, however, is the same as #37.

I've just updated the description in #37 to explain the core of this problem, as well as list some known workarounds. There is also a call to action for any of you in this thread (ping @EmergentBehavior @subvertallchris @jeppebemad @tkroman ) to add scenarios where you are hitting this issue as comments to that ticket. 

This has been a particularly troublesome issue to diagnose and resolve, because the vast majority of the cases where this seemed to be the problem actually were cache poisoning or other low-level concurrency problems. At this point, after much blood has been spilled removing the entire caching subsystem, the core issue outlined in #37 is what remains. We are putting a dedicated team on that issue specifically.

So: In order to help that work, providing us with clear scenarios currently failing on the latest (2.3.1 as of writing this) version of neo that we can convert to integration tests would be very helpful. Please see the description on ticket #37 for details.

Closing this as a duplicate, lets consolidate the discussion in the older ticket.
